src/cascadia/TerminalSettingsEditor/ProfileViewModel.cpp

```diff
@@ -384,9 +402,149 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         BellStyle(currentStyle);
     }
 
+    // Method Description:
+    // - Construct _CurrentBellSounds by importing the _inherited_ value from the model
+    // - Adds a PropertyChanged handler to each BellSoundViewModel to propagate changes to the model
+    void ProfileViewModel::_InitializeCurrentBellSounds()
+    {
+        _CurrentBellSounds = winrt::single_threaded_observable_vector<Editor::BellSoundViewModel>();
+        if (const auto soundList = _profile.BellSound())
+        {
+            for (const auto&& bellSound : soundList)
+            {
+                _CurrentBellSounds.Append(winrt::make<BellSoundViewModel>(bellSound, _dispatcher));
+            }
+        }
+        _MarkDuplicateBellSoundDirectories();
+        _NotifyChanges(L"CurrentBellSounds");
+    }
+
+    // Method Description:
+    // - If the current layer is inheriting the bell sound from its parent,
+    //   we need to copy the _inherited_ bell sound list to the current layer
+    //   so that we can then apply modifications to it
+    void ProfileViewModel::_PrepareModelForBellSoundModification()
+    {
+        if (!_profile.HasBellSound())
+        {
+            std::vector<hstring> newSounds;
+            if (const auto inheritedSounds = _profile.BellSound())
+            {
+                // copy inherited bell sounds to the current layer
+                newSounds.reserve(inheritedSounds.Size());
+                for (const auto sound : inheritedSounds)
+                {
+                    newSounds.push_back(sound);
+                }
+            }
+            // if we didn't inherit any bell sounds,
+            // we should still set the bell sound to an empty list (instead of null)
+            _profile.BellSound(winrt::single_threaded_vector<hstring>(std::move(newSounds)));
+        }
+    }
+
+    // Method Description:
+    // - Check if any bell sounds share the same name.
+    //   If they do, mark them so that they show the directory path in the UI
+    void ProfileViewModel::_MarkDuplicateBellSoundDirectories()
+    {
+        for (uint32_t i = 0; i < _CurrentBellSounds.Size(); i++)
+        {
+            auto soundA = _CurrentBellSounds.GetAt(i);
+            for (uint32_t j = i + 1; j < _CurrentBellSounds.Size(); j++)
+            {
+                auto soundB = _CurrentBellSounds.GetAt(j);
+                if (soundA.DisplayPath() == soundB.DisplayPath())
+                {
+                    get_self<BellSoundViewModel>(_CurrentBellSounds.GetAt(i))->ShowDirectory(true);
+                    get_self<BellSoundViewModel>(_CurrentBellSounds.GetAt(j))->ShowDirectory(true);
+                }
+            }
+        }
+    }
+
+    BellSoundViewModel::BellSoundViewModel(hstring path, const Windows::UI::Core::CoreDispatcher& dispatcher) :
+        _Path{ path },
+        _dispatcher{ dispatcher }
+    {
+        _CheckIfFileExists();

```

Comment:
I'm a bit way of doing this in the constructor, but I guess it's fine..?

Reply:
you will probably need a way to capture a strong reference to `this` so that it's kept alive across the coroutine body

Reply:
is it possible for you to do this all in one coroutine invocation (check all bell sound files) rather than spawning one coroutine per bell sound?

Reply:
Good idea. Moved the validation to be just when we import from the JSON.